Unable to invoke no-args constructor for class org.eclipse.smarthome.core.items.ManagedItemProvider$PersistedItem

When performing some tests for #1000 (after adding a Group via Paper UI) I got the following error. This time the inner class is public and static.

```
15:20:20.493 [96-9c2227ade497] ERROR Events.Framework.org.openhab.core:? - FrameworkEvent ERROR
org.apache.felix.log.LogException: java.lang.RuntimeException: Unable to invoke no-args constructor for class org.eclipse.smarthome.core.items.ManagedItemProvider$PersistedItem. Registering an InstanceCreator with Gson for this type may fix this problem.
	at com.google.gson.internal.ConstructorConstructor$14.construct(ConstructorConstructor.java:226)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:210)
	at com.google.gson.Gson.fromJson(Gson.java:888)
	at com.google.gson.Gson.fromJson(Gson.java:853)
	at com.google.gson.Gson.fromJson(Gson.java:802)
	at com.google.gson.Gson.fromJson(Gson.java:774)
	at org.eclipse.smarthome.storage.mapdb.internal.MapDbStorage.deserialize(MapDbStorage.java:178)
	at org.eclipse.smarthome.storage.mapdb.internal.MapDbStorage.get(MapDbStorage.java:104)
	at org.eclipse.smarthome.core.common.registry.AbstractManagedProvider.lambda$0(AbstractManagedProvider.java:71)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1556)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.eclipse.smarthome.core.common.registry.AbstractManagedProvider.getAll(AbstractManagedProvider.java:77)
	at org.eclipse.smarthome.core.internal.items.ItemRegistryImpl.addProvider(ItemRegistryImpl.java:390)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry$ProviderTracker.addingService(AbstractRegistry.java:128)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry$ProviderTracker.addingService(AbstractRegistry.java:1)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
	at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:903)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:112)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:923)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:233)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:865)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:804)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:130)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:228)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:472)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:906)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:892)
	at org.apache.felix.scr.impl.manager.RegistrationManager.changeRegistration(RegistrationManager.java:128)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.registerService(AbstractComponentManager.java:959)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:732)
	at org.apache.felix.scr.impl.manager.DependencyManager$SingleStaticCustomizer.addedService(DependencyManager.java:1053)
	at org.apache.felix.scr.impl.manager.DependencyManager$SingleStaticCustomizer.addedService(DependencyManager.java:1007)
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1216)
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1137)
	at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.trackAdding(ServiceTracker.java:944)
	at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.track(ServiceTracker.java:880)
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:1168)
	at org.apache.felix.scr.impl.BundleComponentActivator$ListenerInfo.serviceChanged(BundleComponentActivator.java:125)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:112)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:923)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:233)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:865)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:804)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:130)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:228)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:472)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:906)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:892)
	at org.apache.felix.scr.impl.manager.RegistrationManager.changeRegistration(RegistrationManager.java:128)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.registerService(AbstractComponentManager.java:959)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:732)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:666)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:432)
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:665)
	at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:339)
	at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:381)
	at org.apache.felix.scr.impl.Activator.access$200(Activator.java:49)
	at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:263)
	at org.apache.felix.scr.impl.AbstractExtender.createExtension(AbstractExtender.java:196)
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:169)
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:49)
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:488)
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232)
	at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:450)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:911)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:233)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151)
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:233)
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:140)
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:132)
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:194)
	at org.eclipse.osgi.container.Module.publishEvent(Module.java:479)
	at org.eclipse.osgi.container.Module.start(Module.java:470)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:431)
	at aQute.launcher.Launcher.startBundles(Launcher.java:528)
	at aQute.launcher.Launcher.activate(Launcher.java:427)
	at aQute.launcher.Launcher.run(Launcher.java:306)
	at aQute.launcher.Launcher.main(Launcher.java:152)
Caused by: java.lang.UnsupportedOperationException: Cannot allocate class org.eclipse.smarthome.core.items.ManagedItemProvider$PersistedItem
	at com.google.gson.internal.UnsafeAllocator$4.newInstance(UnsafeAllocator.java:104)
	at com.google.gson.internal.ConstructorConstructor$14.construct(ConstructorConstructor.java:223)
	... 88 common frames omitted
```

--- a/bundles/org.openhab.core.storage.mapdb/src/main/java/org/eclipse/smarthome/storage/mapdb/internal/MapDbStorage.java
+++ b/bundles/org.openhab.core.storage.mapdb/src/main/java/org/eclipse/smarthome/storage/mapdb/internal/MapDbStorage.java
@@ -20,6 +20,8 @@ import java.util.Map;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.items.ManagedItemProvider.PersistedItem;
+import org.eclipse.smarthome.core.items.ManagedItemProvider.PersistedItemInstanceCreator;
 import org.eclipse.smarthome.core.storage.DeletableStorage;
 import org.eclipse.smarthome.core.storage.Storage;
 import org.mapdb.DB;
@@ -67,7 +69,8 @@ public class MapDbStorage<T> implements DeletableStorage<T> {
         this.db = db;
         this.classLoader = classLoader;
         this.map = db.createTreeMap(name).makeOrGet();
-        this.mapper = new GsonBuilder().registerTypeAdapterFactory(new PropertiesTypeAdapterFactory()).create();
+        this.mapper = new GsonBuilder().registerTypeAdapterFactory(new PropertiesTypeAdapterFactory())
+                .registerTypeAdapter(PersistedItem.class, new PersistedItemInstanceCreator()).create();
     }
 
     @Override

--- a/bundles/org.openhab.core.storage.mapdb/src/test/java/org/eclipse/smarthome/storage/mapdb/internal/MapDbStorageServiceTest.java
+++ b/bundles/org.openhab.core.storage.mapdb/src/test/java/org/eclipse/smarthome/storage/mapdb/internal/MapDbStorageServiceTest.java
@@ -26,6 +26,7 @@ import java.util.Objects;
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.smarthome.config.core.ConfigConstants;
+import org.eclipse.smarthome.core.library.CoreItemFactory;
 import org.eclipse.smarthome.core.storage.DeletableStorage;
 import org.eclipse.smarthome.core.storage.Storage;
 import org.hamcrest.Matchers;
@@ -174,8 +175,8 @@ public class MapDbStorageServiceTest {
     @Test
     public void serializationDeserialization() {
         Assert.assertThat(storage.getKeys().size(), Matchers.equalTo(0));
-        storage.put(KEY_1, new PersistedItem("String", Arrays.asList("LIGHT", "GROUND_FLOOR")));
-        storage.put(KEY_2, new PersistedItem("Number", Arrays.asList("TEMPERATURE", "OUTSIDE")));
+        storage.put(KEY_1, new PersistedItem(CoreItemFactory.STRING, Arrays.asList("LIGHT", "GROUND_FLOOR")));
+        storage.put(KEY_2, new PersistedItem(CoreItemFactory.NUMBER, Arrays.asList("TEMPERATURE", "OUTSIDE")));
         Assert.assertThat(storage.getKeys().size(), Matchers.equalTo(2));
         final Object persistedObject = storage.get(KEY_1);
         Assert.assertThat(persistedObject, Matchers.instanceOf(PersistedItem.class));
@@ -194,25 +195,26 @@ public class MapDbStorageServiceTest {
 
         Assert.assertEquals(0, storage.getKeys().size());
 
-        persistedObject = storage.put(KEY_1, new PersistedItem("String", Arrays.asList("LIGHT", "GROUND_FLOOR")));
+        persistedObject = storage.put(KEY_1,
+                new PersistedItem(CoreItemFactory.STRING, Arrays.asList("LIGHT", "GROUND_FLOOR")));
         Assert.assertEquals(1, storage.getKeys().size());
         Assert.assertNull(persistedObject);
 
         persistedObject = storage.get(KEY_1);
         Assert.assertTrue(persistedObject instanceof PersistedItem);
         persistedItem = (PersistedItem) persistedObject;
-        Assert.assertEquals("String", persistedItem.itemType);
+        Assert.assertEquals(CoreItemFactory.STRING, persistedItem.itemType);
 
-        persistedObject = storage.put(KEY_1, new PersistedItem("Number", Arrays.asList("TEMPERATURE")));
+        persistedObject = storage.put(KEY_1, new PersistedItem(CoreItemFactory.NUMBER, Arrays.asList("TEMPERATURE")));
         Assert.assertTrue(persistedObject instanceof PersistedItem);
         persistedItem = (PersistedItem) persistedObject;
         Assert.assertEquals(1, storage.getKeys().size());
-        Assert.assertEquals("String", persistedItem.itemType);
+        Assert.assertEquals(CoreItemFactory.STRING, persistedItem.itemType);
 
         persistedObject = storage.get(KEY_1);
         Assert.assertTrue(persistedObject instanceof PersistedItem);
         persistedItem = (PersistedItem) persistedObject;
-        Assert.assertEquals("Number", persistedItem.itemType);
+        Assert.assertEquals(CoreItemFactory.NUMBER, persistedItem.itemType);
 
         storage.remove(KEY_1);
         Assert.assertEquals(0, storage.getKeys().size());

--- a/bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProvider.java
+++ b/bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProvider.java
@@ -12,6 +12,7 @@
  */
 package org.eclipse.smarthome.core.items;
 
+import java.lang.reflect.Type;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -24,7 +25,8 @@ import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CopyOnWriteArrayList;
 
-import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.smarthome.core.common.registry.AbstractManagedProvider;
 import org.eclipse.smarthome.core.items.ManagedItemProvider.PersistedItem;
 import org.eclipse.smarthome.core.items.dto.GroupFunctionDTO;
@@ -37,6 +39,8 @@ import org.osgi.service.component.annotations.ReferenceCardinality;
 import org.osgi.service.component.annotations.ReferencePolicy;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.google.gson.InstanceCreator;
 
 /**
  * {@link ManagedItemProvider} is an OSGi service, that allows to add or remove
@@ -53,38 +57,35 @@ import org.slf4j.LoggerFactory;
 @Component(immediate = true, service = { ItemProvider.class, ManagedItemProvider.class })
 public class ManagedItemProvider extends AbstractManagedProvider<Item, String, PersistedItem> implements ItemProvider {
 
+    @NonNullByDefault
     public static class PersistedItem {
 
-        public PersistedItem(@NonNull String itemType) {
+        public PersistedItem(String itemType) {
             this.itemType = itemType;
         }
 
-        public String baseItemType;
-
-        public List<String> groupNames;
-
-        public @NonNull String itemType;
-
-        public Set<String> tags;
-
-        public String label;
-
-        public String category;
-
-        public String functionName;
-
-        public List<String> functionParams;
-
-        public String dimension;
-
+        public @Nullable String baseItemType;
+        public @Nullable List<String> groupNames;
+        public String itemType;
+        public @Nullable Set<String> tags;
+        public @Nullable String label;
+        public @Nullable String category;
+        public @Nullable String functionName;
+        public @Nullable List<String> functionParams;
+        public @Nullable String dimension;
     }
 
-    private static final @NonNull String ITEM_TYPE_GROUP = "Group";
+    public static class PersistedItemInstanceCreator implements InstanceCreator<PersistedItem> {
+
+        @Override
+        public PersistedItem createInstance(Type type) {
+            return new PersistedItem("");
+        }
+    }
 
     private final Logger logger = LoggerFactory.getLogger(ManagedItemProvider.class);
 
     private final Collection<ItemFactory> itemFactories = new CopyOnWriteArrayList<>();
-
     private final Map<String, PersistedItem> failedToCreate = new ConcurrentHashMap<>();
 
     @Activate
@@ -103,19 +104,19 @@ public class ManagedItemProvider extends AbstractManagedProvider<Item, String, P
         Item item = get(itemName);
         if (recursive && item instanceof GroupItem) {
             for (String member : getMemberNamesRecursively((GroupItem) item, getAll())) {
-                this.remove(member);
+                remove(member);
             }
         }
         if (item != null) {
-            this.remove(item.getName());
+            remove(item.getName());
             return item;
         } else {
             return null;
         }
     }
 
-    private List<@NonNull String> getMemberNamesRecursively(GroupItem groupItem, Collection<Item> allItems) {
-        List<@NonNull String> memberNames = new ArrayList<>();
+    private List<String> getMemberNamesRecursively(GroupItem groupItem, Collection<Item> allItems) {
+        List<String> memberNames = new ArrayList<>();
         for (Item item : allItems) {
             if (item.getGroupNames().contains(groupItem.getName())) {
                 memberNames.add(item.getName());
@@ -127,8 +128,8 @@ public class ManagedItemProvider extends AbstractManagedProvider<Item, String, P
         return memberNames;
     }
 
-    private Item createItem(@NonNull String itemType, @NonNull String itemName) {
-        for (ItemFactory factory : this.itemFactories) {
+    private Item createItem(String itemType, String itemName) {
+        for (ItemFactory factory : itemFactories) {
             Item item = factory.createItem(itemType, itemName);
             if (item != null) {
                 return item;
@@ -147,7 +148,7 @@ public class ManagedItemProvider extends AbstractManagedProvider<Item, String, P
      * @param item the Item to translate the name
      * @return the translated ItemTypeName understandable by the {@link ItemFactory}s
      */
-    private @NonNull String toItemFactoryName(Item item) {
+    private String toItemFactoryName(Item item) {
         return item.getType();
     }
 
@@ -197,7 +198,7 @@ public class ManagedItemProvider extends AbstractManagedProvider<Item, String, P
     protected Item toElement(String itemName, PersistedItem persistedItem) {
         Item item = null;
 
-        if (persistedItem.itemType.equals(ITEM_TYPE_GROUP)) {
+        if (persistedItem.itemType.equals(GroupItem.TYPE)) {
             if (persistedItem.baseItemType != null) {
                 Item baseItem = createItem(persistedItem.baseItemType, itemName);
                 if (persistedItem.functionName != null) {
@@ -259,7 +260,7 @@ public class ManagedItemProvider extends AbstractManagedProvider<Item, String, P
     @Override
     protected PersistedItem toPersistableElement(Item item) {
         PersistedItem persistedItem = new PersistedItem(
-                item instanceof GroupItem ? ITEM_TYPE_GROUP : toItemFactoryName(item));
+                item instanceof GroupItem ? GroupItem.TYPE : toItemFactoryName(item));
 
         if (item instanceof GroupItem) {
             GroupItem groupItem = (GroupItem) item;
@@ -292,5 +293,4 @@ public class ManagedItemProvider extends AbstractManagedProvider<Item, String, P
             }
         }
     }
-
 }

--- a/itests/org.openhab.core.tests/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProviderOSGiTest.java
+++ b/itests/org.openhab.core.tests/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProviderOSGiTest.java
@@ -20,6 +20,7 @@ import java.util.Collections;
 import java.util.List;
 
 import org.eclipse.smarthome.core.items.ManagedItemProvider.PersistedItem;
+import org.eclipse.smarthome.core.library.CoreItemFactory;
 import org.eclipse.smarthome.core.library.items.NumberItem;
 import org.eclipse.smarthome.core.library.items.StringItem;
 import org.eclipse.smarthome.core.library.items.SwitchItem;
@@ -119,7 +120,7 @@ public class ManagedItemProviderOSGiTest extends JavaOSGiTest {
         itemProvider.add(new StringItem("Item"));
         Item result = itemProvider.update(new SwitchItem("Item"));
 
-        assertThat(result.getType(), is("String"));
+        assertThat(result.getType(), is(CoreItemFactory.STRING));
 
         itemProvider.remove("Item");
 

